File: services/orb_dictate_speaker.py
"""ORB Dictate speaker/participant helpers — introductions, not biometrics."""

from __future__ import annotations

import re
from typing import Any
from uuid import uuid4

from schemas.orb_dictate import (
    OrbDictateParticipant,
    OrbDictateSpeakerSummary,
    OrbDictateTranscriptSegment,
)

SPEAKER_BOUNDARY_COPY = (
    "Speaker labels are based on introductions and your corrections. "
    "ORB Dictate does not verify identity by voice."
)

INTRO_PATTERNS: list[tuple[re.Pattern[str], str]] = [
    (re.compile(r"^(.+?),\s*(.+?),\s*speaking\.?$", re.I), "name_role"),
    (re.compile(r"^(.+?)\s+speaking\.?$", re.I), "name_only"),
    (re.compile(r"^this is\s+(.+?)(?:\.|$)", re.I), "name_only"),
    (re.compile(r"^my name is\s+(.+?)(?:\.|$)", re.I), "name_only"),
    (re.compile(r"^i['']?m\s+(.+?)(?:\.|$)", re.I), "name_only"),
    (re.compile(r"^(.+?),\s*(deputy|registered|manager|worker|nurse|social worker).*$", re.I), "name_role_loose"),
]

ROLE_SUFFIX = re.compile(
    r",\s*(registered manager|deputy manager|team manager|support worker|"
    r"residential support worker|social worker|nurse|key worker|manager)\s*$",
    re.I,
)


def _new_id(prefix: str = "p") -> str:
    return f"{prefix}_{uuid4().hex[:10]}"


def _new_segment_id() -> str:
    return f"seg_{uuid4().hex[:10]}"


def parse_introduction_line(line: str) -> tuple[str, str | None] | None:
    text = line.strip().strip('"').strip("'")
    if not text or len(text) > 200:
        return None
    for pattern, kind in INTRO_PATTERNS:
        m = pattern.match(text)
        if not m:
            continue
        if kind == "name_role":
            return m.group(1).strip(), m.group(2).strip()
        if kind == "name_role_loose":
            name = m.group(1).strip()
            role_match = ROLE_SUFFIX.search(name)
            if role_match:
                return name[: role_match.start()].strip(), role_match.group(1).strip()
            return name, m.group(2).strip() if m.lastindex and m.lastindex >= 2 else None
        return m.group(1).strip(), None
    return None


def suggest_participants_from_text(text: str) -> list[OrbDictateParticipant]:
    suggestions: list[OrbDictateParticipant] = []
    seen: set[str] = set()
    for line in text.splitlines():
        parsed = parse_introduction_line(line)
        if not parsed:
            continue
        name, role = parsed
        key = name.lower()
        if key in seen:
            continue
        seen.add(key)
        suggestions.append(
            OrbDictateParticipant(
                id=_new_id(),
                name=name,
                role=role,
                introduced_by="self",
            )
        )
    return suggestions


def participant_label(participant: OrbDictateParticipant) -> str:
    if participant.role:
        return f"{participant.name}, {participant.role}"
    return participant.name


def _normalise_segment_source(source: str) -> str:
    mapping = {
        "dictation": "live",
        "template": "paste",
        "live": "live",
        "upload": "upload",
        "paste": "paste",
        "orb_voice": "orb_voice",
    }
    return mapping.get(source, "paste")


def text_to_segments(
    text: str,
    *,
    source: str = "paste",
    participants: list[OrbDictateParticipant] | None = None,
) -> list[OrbDictateTranscriptSegment]:
    source = _normalise_segment_source(source)
    """Split transcript into segments; detect inline speaker labels like 'Tom Kelly:'."""
    participants = participants or []
    by_name = {p.name.lower(): p for p in participants}
    segments: list[OrbDictateTranscriptSegment] = []
    label_re = re.compile(r"^([A-Z][A-Za-z' -]{1,60})(?:,\s*[^:\n]{1,80})?:\s*(.+)$")
    unknown_idx = 0

    for block in re.split(r"\n\s*\n", text.strip()):
        block = block.strip()
        if not block:
            continue
        speaker_id: str | None = None
        speaker_label = "Speaker 1"
        body = block

        first_line, _, rest = block.partition("\n")
        m = label_re.match(first_line.strip())
        if m:
            name = m.group(1).strip()
            body = m.group(2).strip()
            if rest:
                body = f"{body}\n{rest}".strip()
            p = by_name.get(name.lower())
            if p:
                speaker_id = p.id
                speaker_label = participant_label(p)
            else:
                unknown_idx += 1
                speaker_label = name if "," not in name else name
        else:
            intro = parse_introduction_line(block.split("\n")[0])
            if intro and len(block.split()) < 12:
                continue
            unknown_idx = max(unknown_idx, 0) + 1
            speaker_label = f"Speaker {unknown_idx}" if unknown_idx else "Speaker 1"

        segments.append(
            OrbDictateTranscriptSegment(
                id=_new_segment_id(),
                speaker_id=speaker_id,
                speaker_label=speaker_label,
                text=body,
                source=source,  # type: ignore[arg-type]
            )
        )
    if not segments and text.strip():
        segments.append(
            OrbDictateTranscriptSegment(
                id=_new_segment_id(),
                speaker_label="Speaker 1",
                text=text.strip(),
                source=source,  # type: ignore[arg-type]
            )
        )
    return segments


def segments_to_plain_text(segments: list[OrbDictateTranscriptSegment]) -> str:
    lines: list[str] = []
    for seg in segments:
        label = seg.speaker_label or "Speaker"
        lines.append(f"{label}: {seg.text.strip()}")
    return "\n\n".join(lines)


def build_speaker_summary(
    participants: list[OrbDictateParticipant],
    segments: list[OrbDictateTranscriptSegment],
) -> OrbDictateSpeakerSummary:
    known_ids = {p.id for p in participants}
    labels_in_segments = {s.speaker_label for s in segments if s.speaker_label}
    known_speakers = sum(
        1
        for p in participants
        if p.id in {s.speaker_id for s in segments if s.speaker_id}
        or p.name in labels_in_segments
        or participant_label(p) in labels_in_segments
    )
    unknown_labels = [
        s.speaker_label
        for s in segments
        if s.speaker_label.startswith("Speaker ") and not s.speaker_id
    ]
    unknown_speakers = len(set(unknown_labels))
    needs_review = bool(unknown_speakers) or any(
        s.speaker_id not in known_ids and s.speaker_id for s in segments
    )
    return OrbDictateSpeakerSummary(
        known_speakers=known_speakers,
        unknown_speakers=unknown_speakers,
        needs_review=needs_review or len(participants) == 0,
    )


def anonymise_note_text(
    text: str,
    participants: list[OrbDictateParticipant],
) -> str:
    result = text
    for p in participants:
        if p.name:
            replacement = p.role or "Staff member"
            result = re.sub(re.escape(p.name), replacement, result, flags=re.I)
    return result


def participants_block_for_prompt(
    participants: list[OrbDictateParticipant],
    segments: list[OrbDictateTranscriptSegment],
) -> str:
    if not participants and not segments:
        return ""
    lines = ["Participants:"]
    for p in participants:
        role = f" ({p.role})" if p.role else ""
        lines.append(f"- {p.name}{role}")
    if segments:
        lines.append("\nTranscript by speaker:")
        for seg in segments[:80]:
            flag = " [direct quote]" if seg.is_direct_quote else ""
            review = " [needs review]" if seg.needs_review else ""
            lines.append(f"{seg.speaker_label}{flag}{review}: {seg.text[:2000]}")
    return "\n".join(lines)

File: frontend-next/lib/orb/dictate/orb-dictate-speaker.ts
export type OrbDictateParticipantIntroducedBy = 'self' | 'manual' | 'import' | 'unknown'

export type OrbDictateParticipant = {
  id: string
  name: string
  role?: string
  organisation?: string
  initials?: string
  introducedBy?: OrbDictateParticipantIntroducedBy
}

export type OrbDictateSegmentSource = 'live' | 'upload' | 'paste' | 'orb_voice'

export type OrbDictateTranscriptSegment = {
  id: string
  speaker_id?: string
  speaker_label: string
  text: string
  started_at?: string
  ended_at?: string
  confidence?: number
  source: OrbDictateSegmentSource
  is_direct_quote?: boolean
  needs_review?: boolean
}

export type OrbDictateSpeakerSummary = {
  known_speakers: number
  unknown_speakers: number
  needs_review: boolean
}

export type OrbDictateMode =
  | 'rough_note'
  | 'team_meeting'
  | 'staff_debrief'
  | 'investigation_meeting'
  | 'reflective_supervision'
  | 'strategy_multi_agency_prep'
  | 'handover'

export const ORB_DICTATE_MODE_LABELS: Record<OrbDictateMode, string> = {
  rough_note: 'Rough note',
  team_meeting: 'Team meeting',
  staff_debrief: 'Staff debrief',
  investigation_meeting: 'Investigation meeting',
  reflective_supervision: 'Reflective supervision',
  strategy_multi_agency_prep: 'Strategy / multi-agency prep',
  handover: 'Handover'
}

export const MODE_REQUIRES_CONSENT: OrbDictateMode[] = [
  'team_meeting',
  'staff_debrief',
  'investigation_meeting',
  'reflective_supervision',
  'strategy_multi_agency_prep',
  'handover'
]

export const SPEAKER_BOUNDARY_COPY =
  'Speaker labels are based on introductions and your corrections. ORB Dictate does not verify identity by voice.'

export const SPEAKER_INTRO_PROMPT =
  "Before you start, you can introduce who is present. For example: 'Tom Kelly, Registered Manager, speaking.' ORB will use this to label the transcript."

function newId(prefix: string) {
  return `${prefix}_${Math.random().toString(36).slice(2, 11)}`
}

const INTRO_PATTERNS: Array<{ pattern: RegExp; kind: string }> = [
  { pattern: /^(.+?),\s*(.+?),\s*speaking\.?$/i, kind: 'name_role' },
  { pattern: /^(.+?)\s+speaking\.?$/i, kind: 'name_only' },
  { pattern: /^this is\s+(.+?)(?:\.|$)/i, kind: 'name_only' },
  { pattern: /^my name is\s+(.+?)(?:\.|$)/i, kind: 'name_only' },
  { pattern: /^i['']?m\s+(.+?)(?:\.|$)/i, kind: 'name_only' }
]

export function parseIntroductionLine(line: string): { name: string; role?: string } | null {
  const text = line.trim().replace(/^["']|["']$/g, '')
  if (!text || text.length > 200) return null
  for (const { pattern, kind } of INTRO_PATTERNS) {
    const m = text.match(pattern)
    if (!m) continue
    if (kind === 'name_role') return { name: m[1].trim(), role: m[2].trim() }
    return { name: m[1].trim() }
  }
  return null
}

export function suggestParticipantsFromText(text: string): OrbDictateParticipant[] {
  const suggestions: OrbDictateParticipant[] = []
  const seen = new Set<string>()
  for (const line of text.split('\n')) {
    const parsed = parseIntroductionLine(line)
    if (!parsed) continue
    const key = parsed.name.toLowerCase()
    if (seen.has(key)) continue
    seen.add(key)
    suggestions.push({
      id: newId('p'),
      name: parsed.name,
      role: parsed.role,
      introducedBy: 'self'
    })
  }
  return suggestions
}

export function participantLabel(p: OrbDictateParticipant): string {
  return p.role ? `${p.name}, ${p.role}` : p.name
}

export function textToSegments(
  text: string,
  source: OrbDictateSegmentSource = 'paste',
  participants: OrbDictateParticipant[] = []
): OrbDictateTranscriptSegment[] {
  const byName = new Map(participants.map((p) => [p.name.toLowerCase(), p]))
  const segments: OrbDictateTranscriptSegment[] = []
  const labelRe = /^([A-Z][A-Za-z' -]{1,60})(?:,\s*[^:\n]{1,80})?:\s*(.+)$/
  let unknownIdx = 0

  const blocks = text.trim().split(/\n\s*\n/)
  for (const block of blocks) {
    const trimmed = block.trim()
    if (!trimmed) continue
    let speakerId: string | undefined
    let speakerLabel = 'Speaker 1'
    let body = trimmed

    const firstLine = trimmed.split('\n')[0]
    const rest = trimmed.slice(firstLine.length).trim()
    const m = firstLine.match(labelRe)
    if (m) {
      const name = m[1].trim()
      body = m[2].trim()
      if (rest) body = `${body}\n${rest}`.trim()
      const p = byName.get(name.toLowerCase())
      if (p) {
        speakerId = p.id
        speakerLabel = participantLabel(p)
      } else {
        unknownIdx += 1
        speakerLabel = name
      }
    } else {
      const intro = parseIntroductionLine(firstLine)
      if (intro && trimmed.split(/\s+/).length < 12) continue
      unknownIdx += 1
      speakerLabel = `Speaker ${unknownIdx}`
    }

    segments.push({
      id: newId('seg'),
      speaker_id: speakerId,
      speaker_label: speakerLabel,
      text: body,
      source
    })
  }

  if (!segments.length && text.trim()) {
    segments.push({
      id: newId('seg'),
      speaker_label: 'Speaker 1',
      text: text.trim(),
      source
    })
  }
  return segments
}

export function segmentsToPlainText(segments: OrbDictateTranscriptSegment[]): string {
  return segments
    .map((s) => `${s.speaker_label}: ${s.text.trim()}`)
    .join('\n\n')
}

export function buildSpeakerSummary(
  participants: OrbDictateParticipant[],
  segments: OrbDictateTranscriptSegment[]
): OrbDictateSpeakerSummary {
  const labels = new Set(segments.map((s) => s.speaker_label))
  const known = participants.filter(
    (p) =>
      segments.some((s) => s.speaker_id === p.id) ||
      labels.has(p.name) ||
      labels.has(participantLabel(p))
  ).length
  const unknownLabels = new Set(
    segments.filter((s) => /^Speaker \d+$/i.test(s.speaker_label) && !s.speaker_id).map((s) => s.speaker_label)
  )
  return {
    known_speakers: known,
    unknown_speakers: unknownLabels.size,
    needs_review: unknownLabels.size > 0 || participants.length === 0
  }
}

export function anonymiseText(text: string, participants: OrbDictateParticipant[]): string {
  let result = text
  for (const p of participants) {
    if (!p.name) continue
    const replacement = p.role || 'Staff member'
    result = result.replace(new RegExp(escapeRegExp(p.name), 'gi'), replacement)
  }
  return result
}

function escapeRegExp(value: string) {
  return value.replace(/[.*+?^${}()|[\]\\]/g, '\\$&')
}

export type OrbDictateNoteTypeForMode =
  | 'daily_record'
  | 'team_meeting'
  | 'staff_debrief'
  | 'investigation_meeting'
  | 'supervision_reflection'
  | 'strategy_multi_agency_prep'
  | 'handover_note'

export function modeToNoteType(mode: OrbDictateMode): OrbDictateNoteTypeForMode {
  const map: Record<OrbDictateMode, OrbDictateNoteTypeForMode> = {
    rough_note: 'daily_record',
    team_meeting: 'team_meeting',
    staff_debrief: 'staff_debrief',
    investigation_meeting: 'investigation_meeting',
    reflective_supervision: 'supervision_reflection',
    strategy_multi_agency_prep: 'strategy_multi_agency_prep',
    handover: 'handover_note'
  }
  return map[mode]
}

export function voiceTurnsToSegments(
  turns: Array<{ role: string; text: string }>
): OrbDictateTranscriptSegment[] {
  let speakerNum = 0
  return turns
    .filter((t) => t.role === 'user' || t.role === 'assistant')
    .map((t) => {
      if (t.role === 'user') speakerNum += 1
      const label = t.role === 'assistant' ? 'ORB' : `Speaker ${speakerNum || 1}`
      return {
        id: newId('seg'),
        speaker_label: label,
        text: t.text,
        source: 'orb_voice' as const,
        needs_review: t.role === 'user'
      }
    })
}
